updated instructions for adding a new rpm (#76)
open-edge-platform/edge-microvisor-toolkit: docs/developer-guide/get-started/building-howto.md

will include the set of all `rpms` specified within the array of `packagelist` files from the
`imageconfig`.

> ### Example: Adding an existing rpm (Nano)

Reviewer: In all cases where RPM is used to reference the general file type, I recommend using all caps `RPMs` or `RPM packages`, and `.rpm` in references to specific files (`hello-world.rpm`)

Reviewer 2: ```suggestion
### Example: Adding an existing RPM (Nano)
```

The following example shows how to add `nano` as an alternative text editor to the image.

> You can add the packages for which `.spec` files already exist. Simply include them in an

Reviewer: In all cases where SPEC is used to reference the general file type, I recommend using all caps SPECs or SPEC files, and .spec in references to specific files (`hello-world.spec`)

```bash
sudo make image -j8 REBUILD_TOOLS=y REBUILD_PACKAGES=n CONFIG_FILE=./imageconfigs/edge-image.json
```
### Add a new package

To add a new package you need to generate a `SPEC` file for the package which
contains all required information for the build infrastructure to generate the
`SRPM` and `RPM` for the package. There are a few steps involved to create

> a new package for the EMT.

Reviewer 3: ```suggestion
a new package for Edge Microvisor Toolkit.
```

1. Defining the SPEC file and add it into the `/SPECS` directory

> 1. Create the source archive and generate the sha256sum for the package

Reviewer 3: ```suggestion
1. Creating the source archive and generating the sha256sum for the package
```
to be consistent with rest of points

> 1. Uploading the tar.gz package to the source package repository

Reviewer 3: I think we should move this step to the bottom after package has been tested locally.

> 1. Add the SPEC file /SPEC directory and updating the `cgmanifest.json` file

Reviewer 3: ```suggestion
1. Add the SPEC file in /SPECS directory and updating the `cgmanifest.json` file
```

> 

Reviewer 3: Can we add one more step here to say generate an image with this package included and test it?

%prep
%setup -q

%build
# Nothing to build for a shell script

%install
mkdir -p %{buildroot}/usr/bin
install -m 0755 helloworld.sh %{buildroot}/usr/bin/helloworld

mkdir -p %{buildroot}/usr/share/helloworld
install -m 0644 helloworld.signature.json %{buildroot}/usr/share/helloworld/

%files
/usr/bin/helloworld
/usr/share/helloworld/helloworld.signature.json

%changelog
* Wed May 01 2025 Your Name <you@example.com> - 1.0-1
- Initial package
```

**Create the archive**: Create your source archive, add the simple script and
make it executable.

```bash
mkdir helloworld-1.0

cat <<EOF > ./helloworld-1.0/helloworld.sh
#!/bin/bash
echo "Hello, world!"
EOF

chmod +x ./helloworld-1.0/helloworld.sh

sha256sum helloworld-1.0/helloworld.sh | awk '{print $1}' > sum.txt
read sum < sum.txt
echo "{\"file\": \"helloworld.sh\", \"sha256\": \"$sum\"}" > helloworld-1.0/helloworld.signature.json

tar -czf helloworld-1.0.tar.gz helloworld-1.0/
```

Copy and build the `rpm` package to the rpm building directory.

```bash
cp helloworld-1.0.tar.gz ~/rpmbuild/SOURCES/
cp helloworld.spec ~/rpmbuild/SPECS/
rpmbuild -ba ~/rpmbuild/SPECS/helloworld.spec
```

> **Uploading the archive**: Upload the tar.gz archive to [TBD].

Reviewer 3: Can we move this to the bottom after image generation and testing? Add something like:

If testing is complete and you are ready to contribute this package, please raise a PR and work with a code owner to upload the source tarball to package source mirror.

**Adding the SPEC**: Add the `helloworld.spec` and the 'helloworld.spec.signature`
file to the `/SPECS` directory. Finally update the `cgmanifest` by using the
provided `python` script.

```bash
    python3 -m pip install -r ./toolkit/scripts/requirements.txt
    python3 ./toolkit/scripts/update_cgmanifest.py first cgmanifest.json ./SPECS/helloworld.spec

> ```

Reviewer 3: After this, should we add steps to:
- Build this using `make build-packages`
- optionally add this as part of the image and test (link to that documentation)?

```bash
sudo make image -j8 REBUILD_TOOLS=y REBUILD_PACKAGES=n CONFIG_FILE=./imageconfigs/edge-image.json
```
### Add a new package

To add a new package you need to generate a `SPEC` file for the package which
contains all required information for the build infrastructure to generate the
`SRPM` and `RPM` for the package. There are a few steps involved to create
a new package for Edge Microvisor Toolkit.

> 1. Defining the SPEC file and add it into the `/SPECS` directory

Reviewer 4: We first need to create a folder with same name as spec file under SPECS/ directory. That should be step 1.

1. Creating the source archive and generating the sha256sum for the package

> 1. Add the SPEC file in /SPECS directory and updating the `cgmanifest.json` file

Reviewer 4: This step and step 1 are similar.

1. Build an image with the package included and test locally
1. Uploading the tar.gz package to the source package repository after is has been tested locally

You need to first install the required build tools for `rpm`. On Fedora you
can simply install the required packages with:

```bash

> sudo dnf install rpm-build rpmdevtools

Reviewer 4: Why do we need to install this separately? I think this comes already when installing tool chain initially.

Reviewer 3: I think it's good to be explicit when possible

Reviewer 3: > Why do we need to install this separately? I think this comes already when installing tool chain initially.

I think it's good to be explicit when possible

echo '%_topdir %(echo $HOME)/rpmbuild' > ~/.rpmmacros
```

**Defining the SPEC file**
Define and test that the SPEC file works as expected and is locally generating
the required artifacts. This example builds a simple hello world `rpm` package
which contains a bash scripts that prints hello world.

```bash
Name:           helloworld
Version:        1.0
Release:        1%{?dist}
Summary:        Simple Hello World script

License:        MIT
URL:            https://example.com/helloworld
Source0:        helloworld-1.0.tar.gz

BuildArch:      noarch

%description
A very basic "Hello World" script packaged as an RPM.

%prep
%setup -q

%build
# Nothing to build for a shell script

%install
mkdir -p %{buildroot}/usr/bin
install -m 0755 helloworld.sh %{buildroot}/usr/bin/helloworld

mkdir -p %{buildroot}/usr/share/helloworld
install -m 0644 helloworld.signature.json %{buildroot}/usr/share/helloworld/

%files
/usr/bin/helloworld
/usr/share/helloworld/helloworld.signature.json

%changelog
* Wed May 01 2025 Your Name <you@example.com> - 1.0-1
- Initial package
```

**Create the archive**: Create your source archive, add the simple script and
make it executable.

```bash

> mkdir helloworld-1.0

Reviewer 4: Inside which folder are we doing these steps?

Author: can be anywhere as this is just to build and test it. I have updated to create it under the home folder.

cat <<EOF > ./helloworld-1.0/helloworld.sh
#!/bin/bash
echo "Hello, world!"
EOF

chmod +x ./helloworld-1.0/helloworld.sh

> sha256sum helloworld-1.0/helloworld.sh | awk '{print $1}' > sum.txt

Reviewer 4: Checksum can be auto generated using SRPM_FILE_SIGNATURE_HANDLING=update flag while building image?

Reviewer 3: Both should work

read sum < sum.txt

> echo "{\"file\": \"helloworld.sh\", \"sha256\": \"$sum\"}" > helloworld-1.0/helloworld.signature.json

Reviewer 4: Is this checksum suppose to be for the source file or the tar.gz file?

Author: fixed

tar -czf helloworld-1.0.tar.gz helloworld-1.0/
```

Copy and build the `rpm` package to the rpm building directory.

```bash
cp helloworld-1.0.tar.gz ~/rpmbuild/SOURCES/
cp helloworld.spec ~/rpmbuild/SPECS/
rpmbuild -ba ~/rpmbuild/SPECS/helloworld.spec
```

**Adding the SPEC**: Add the `helloworld.spec` and the 'helloworld.spec.signature`

> file to the `/SPECS` directory. Finally update the `cgmanifest` by using the

Reviewer 4: We need to first create folder named helloworld inside SPECS/ directory.

Author: documented above

```bash
sudo make image -j8 REBUILD_TOOLS=y REBUILD_PACKAGES=n CONFIG_FILE=./imageconfigs/edge-image.json
```
### Add a new package

To add a new package you need to generate a `SPEC` file for the package which
contains all required information for the build infrastructure to generate the

> `SRPM` and `RPM` for the package. There are a few steps involved to create

Reviewer: `involved in creating` - grammar correction

> `SRPM` and `RPM` for the package. There are a few steps involved to create

Reviewer 2: ```suggestion
`SRPM` and `RPM` for the package. There are a few steps involved in creating
```

a new package for Edge Microvisor Toolkit.

> 1. Create a folder, defining the SPEC file and add it into the `/SPECS` directory.

Reviewer 2: ```suggestion
1. Create a folder, define the SPEC file and add it into the `/SPECS` directory.
```

> 2. Create the source archive and generating the sha256sum for the package.

Reviewer 2: ```suggestion
2. Create the source archive and generate the sha256sum for the package.
```

1. Updating the `cgmanifest.json` file
1. Build an image with the package included and test locally

> 1. Uploading the tar.gz package to the source package repository after is has been tested locally

Reviewer: `Create/Update/Build/Upload` in all steps, for consistent sentence structure in all list items. 
Period at the end of all list items.

> 1. Creating a folder, defining the SPEC file and add it into the `/SPECS` directory
> 1. Creating the source archive and generating the sha256sum for the package
> 1. Updating the `cgmanifest.json` file
> 1. Build an image with the package included and test locally
> 1. Uploading the tar.gz package to the source package repository after is has been tested locally

Reviewer 2: ```suggestion
1. Create a folder, defining the SPEC file and add it into the `/SPECS` directory.
2. Create the source archive and generating the sha256sum for the package.
3. Update the `cgmanifest.json` file.
4. Build an image with the package included and test locally.
5. Upload the tar.gz package to the source package repository after is has been tested locally.
```

You need to first install the required build tools for `rpm`. On Fedora you
can simply install the required packages with:

```bash
sudo dnf install rpm-build rpmdevtools
rpmdev-setuptree
```

`rpmdev-setuptree` creates the necessary directories, which you may need to
manually create on an Ubuntu distribution.

```bash
sudo apt-get install rpm
mkdir -p ~/rpmbuild/{BUILD,RPMS,SOURCES,SPECS,SRPMS}
echo '%_topdir %(echo $HOME)/rpmbuild' > ~/.rpmmacros
```

**Defining the SPEC file**

> Define and test that the SPEC file works as expected and is locally generating

Reviewer: Rephrase for clarity:
`Define the SPEC file and test that it works as expected and generates locally the required artifacts.`

Reviewer 2: ```suggestion
Define the SPEC file and test that it works as expected and generates locally the required artifacts.
```

Source0:        helloworld-1.0.tar.gz

BuildArch:      noarch

%description
A very basic "Hello World" script packaged as an RPM.

%prep
%setup -q

%build
# Nothing to build for a shell script

%install
mkdir -p %{buildroot}/usr/bin
install -m 0755 helloworld.sh %{buildroot}/usr/bin/helloworld

mkdir -p %{buildroot}/usr/share/helloworld
install -m 0644 helloworld.signature.json %{buildroot}/usr/share/helloworld/

%files
/usr/bin/helloworld
/usr/share/helloworld/helloworld.signature.json

%changelog
* Wed May 01 2025 Your Name <you@example.com> - 1.0-1
- Initial package
```

**Create the archive**: Create your source archive, add the simple script and
make it executable.

```bash
# 1. Make your source tree and tarball
mkdir -p ~/helloworld-1.0
cat > ~/helloworld-1.0/helloworld.sh <<'EOF'
#!/bin/bash
echo "Hello, world!"
EOF
chmod +x ~/helloworld-1.0/helloworld.sh

tar -czf helloworld-1.0.tar.gz helloworld-1.0/

# 2. Compute its SHA-256
sum=$(sha256sum helloworld-1.0.tar.gz | awk '{print $1}')

# 3. Write the JSON signature for the tarball
cat > helloworld-1.0.tar.gz.signature.json <<EOF
{
  "file": "helloworld-1.0.tar.gz",
  "sha256": "$sum"
}
EOF

```

> Copy and build the `rpm` package to the rpm building directory.

Reviewer 2: ```suggestion
Copy the RPM package to the building directory and build it.
```

```bash
cp helloworld-1.0.tar.gz ~/rpmbuild/SOURCES/
cp helloworld.spec ~/rpmbuild/SPECS/
rpmbuild -ba ~/rpmbuild/SPECS/helloworld.spec
```

**Adding the SPEC**: Add the `helloworld.spec` and the 'helloworld.spec.signature`
file to the `/SPECS` directory. Finally update the `cgmanifest` by using the
provided `python` script.

```bash
    python3 -m pip install -r ./toolkit/scripts/requirements.txt
    python3 ./toolkit/scripts/update_cgmanifest.py first cgmanifest.json ./SPECS/helloworld.spec
```

**Local Build and Testing**:  If testing is complete and you are ready to contribute this package,
please raise a PR and work with a code owner to upload the source tarball to package source mirror.

```bash
make build-packages # to rebuild the packages
```
Follow the steps under [Customizing an image](./building-howto.md#Customizing-an-Image) to create

> an image with the your new package.

Reviewer: `the your new package` -> `your new package`, redundant "the".

Reviewer 2: ```suggestion
an image with your new package.
```
